### DISCUSSION TITLE
MVT, regionMapping, and bin calculations?

### DISCUSSION THREAD

**Discussion by Joebjammin** [TerriaJS/terriajs#6397] [Q&A] [answered]
Hi Folks,
I have been banging my head trying to get region mapping going. I have some progress, but i am having difficulty translating version 7 to 8. I am getting close, and would appreciate a hint. Signed....sore head.

<img width="1439" alt="Result" src="https://user-images.githubusercontent.com/42526991/177705255-4b3903f6-0154-40b8-adfa-c569da8d4aa5.png">

Catalogue File:
{
    "homeCamera": {
      "north": 31.0,
      "east": -80.0,
      "south": 24.5,
      "west": -87.6
    },
    "catalog": [
     {
        "id": "1",
        "type": "geojson",
        "name": "FL_ZCTA2018",
        "url": "test/FL_ZCTA2018.geojson"
      },
      {
        "id": "2",
        "name": "FL_COUNTY",
        "type": "mvt",
        "url": "test/FL_COUNTY/{z}/{x}/{y}.pbf",
        "layer": "FL_COUNTY",
        "minimumZoom": 1,
        "maximumZoom": 10,
        "fillColor": "orange",
        "lineColor": "yellow",
        "rectangle": {
          "west": -87.6348959282059781,
          "south": 24.5210316754366957,
          "east": -80.0313186453441432,
          "north": 31.0001954218924993
        }
      },
      {  
      "id": "3",
      "name": "HEALTH INSURANCE COVERAGE", 
      "type": "csv", 
      "url": "data/HEALTH INSURANCE COVERAGE  - DP03 FLC 2019.csv",
      "description": "",
      "opacity": ".8",
      "tableStyle": {
          "regionColumn": "GEOID",
          //"regionVariable": "GEOID",
          //"regionType":"FLO_COUNTY",
          "numberOfBins": 5,
          "colorPalette": "5-class Oranges",
          "colorBinMethod": "ckmeans" //<<<----------------Is this still supported?  
      }
    }
    ],
    "viewerMode": "3dSmooth",
    "baseMaps": {
      "defaultBaseMapId": "basemap-positron",
      "previewBaseMapId": "basemap-positron"
    }
  }
  
MVT Tiles:
[FL_COUNTY.zip](https://github.com/TerriaJS/terriajs/files/9060914/FL_COUNTY.zip)

Region Mapping:
{
    "regionWmsMap": {
        "FL_COUNTY": {
            "layerName": "FL_COUNTY",
            "server": "data/FL_COUNTY/{z}/{x}/{y}.pbf",
            "serverType": "MVT",
            "serverMinZoom": 0,
            "serverMaxZoom": 10,
            "serverMaxNativeZoom": 9,
            "regionIdsFile": "data/regionids/region_map-FL_COUNTY.json",
            "regionProp": "GEOID",
            "nameProp": "County",
            "aliases": [
                "GEOID",
                "FL_COUNTY"
            ],
            "description": "FLORIDA COUNTIES",
            "bbox": [
                -87.63489593,
                24.52103168,
                -80.03131865,
                31.00019542
            ]
        }
    }
}
Data File:
[HEALTH INSURANCE COVERAGE  - DP03 FLC 2019.csv](https://github.com/TerriaJS/terriajs/files/9060965/HEALTH.INSURANCE.COVERAGE.-.DP03.FLC.2019.csv)

**Comment by nf-s** [comment, marked as answer]
Hi @Joebjammin 

Some of those properties in your catalog aren't supported in v8 - `tableStyle` is quite different now

We have a catalog converter which can help to translate those old properties - https://catalog-converter.terria.io/

At the moment we don't have binning methods implemented - we have an issue here #4228 

We also don't support those "color brewer" strings (`"5-class Oranges"`) - issue #5215

So you need to split it into
- `"numberOfBins": 5`
- `"colorPalette": "Oranges"`

This config works for me

```json
{
  "id": "3",
  "name": "HEALTH INSURANCE COVERAGE",
  "type": "csv",
  "url": "data/HEALTH INSURANCE COVERAGE - DP03 FLC 2019.csv",
  "description": "",
  "opacity": 0.8,
  "defaultStyle": {
    "regionColumn": "GEOID",
    "color": {
      "numberOfBins": 5,
      "colorPalette": "Oranges"
    }
  }
}
```

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/6187649/177724935-b05e9a0a-cc32-4927-8bed-d9996877e6ed.png">

Let me know how you go 🙂 

You didn't supply a regions ID file - so I created this one and it seems to work

```json
{"layer":"region_map:FL_COUNTY","property":"GEOID","values":[12001,12003,12005,12007,12009,12011,12013,12015,12017,12019,12021,12023,12027,12029,12031,12033,12035,12037,12039,12041,12043,12045,12047,12049,12051,12053,12055,12057,12059,12061,12063,12065,12067,12069,12071,12073,12075,12077,12079,12081,12083,12085,12086,12087,12089,12091,12093,12095,12097,12099,12101,12103,12105,12107,12109,12111,12113,12115,12117,12119,12121,12123,12125,12127,12129,12131,12133]}
```

**Reply by Joebjammin** [reply]
Nick, Thank you very much. What binning method is being used? Is there a way to hard code ckmeans method and rebuild?

**Reply by nf-s** [reply]
Current method is equal interval.

You can hard code bins using `binMaximums` - for example

http://ci.terria.io/main/#clean&https://gist.githubusercontent.com/nf-s/11d3f65450eabbcd38371ab15e84cee7/raw/41a6edafe71da665f8dcaa2dcc1ab52ad3f1f182/color-bin-fix.json&share=s-3YMSXGF73TczYKe0nkyYXJOGH04

uses the following catalog JSON

```json
{
      "type": "csv",
      "name": "test",
      "url": " http://ci.terria.io/main/build/TerriaJS/test/csv/youth-unemployment-rate-2018.csv",
      "styles": [
        {
          "id": "youth unemployment (%)",
          "color": {
            "binMaximums": [
              8,
              10,
              15,
              20,
              30,
              50
            ],
            "colorPalette": "PiYG"
          }
        }
      ]
    }
```

Which defines the following bins:

- minimum value to **8**
- 8 to **10**
- 10 to **15**
- 15 to **20**
- 20 to **30**
- 30 to **50**

**Reply by nf-s** [reply]
We also have the new "Edit Style" function - so you can edit the bins in TerriaJS

- Click "Edit Style"
  - <img width="400" alt="image" src="https://user-images.githubusercontent.com/6187649/172037496-ee76ca2c-d275-4247-be0d-69c181e71e1f.png">
- Set style to your liking
- Click "show advanced options"
  - <img width="400" alt="image" src="https://user-images.githubusercontent.com/6187649/172037527-eb76afec-517f-4877-bd5a-dc946f982089.png">
- Click "Copy user stratum to clipboard"
  - <img width="400" alt="image" src="https://user-images.githubusercontent.com/6187649/172037547-5471f26b-2e55-41dd-a9d9-0a3c89835807.png">
- Merge copied JSON into your catalog item JSON

**Note** this feature is brand new and experimental - so you may run into bugs/quirks

**Reply by Joebjammin** [reply]
Thank you very much, Nick!

**Reply by nf-s** [reply]
No problem!

Let us know if you have any other issues 🙂

**Comment by jktittalom** [comment]
Hello @nf-s , I have tried your suggestion but it doesn't work for me.
![image](https://user-images.githubusercontent.com/1362472/178664603-e1f40075-1902-4e5b-a28a-032ef1ad456d.png)

I have these:

Region Map:
{
    "regionWmsMap": {
        "FL_COUNTY": {
            "layerName": "FL_COUNTY",
            "server": "data/s3/FL_COUNTY/{z}/{x}/{y}.pbf",
            "serverType": "MVT",
            "serverMinZoom": 0,
            "serverMaxZoom": 10,
            "serverMaxNativeZoom": 9,

            "regionIdsFile": "data/regionids/region_map-FL_COUNTY.json",
            "regionProp": "GEOID",
            "nameProp": "County",
            "aliases": [
                "GEOID",
                "FL_COUNTY"
            ],
            "description": "FLORIDA Counties",
            "bbox": [
                -87.63489593,
                24.52103168,
                -80.03131865,
                31.00019542
            ]
        }
    }
}

Catalogue File:
{
    "homeCamera": {
       "north": 31,
        "east": -79.72,
        "south": 24.4,
        "west": -87.63
      },
    "catalog": [  
      {
          "id": "3",
          "name": "HEALTH INSURANCE COVERAGE",
          "type": "csv",
          "url": "data/HEALTH INSURANCE COVERAGE  - DP03 FLC 2019.csv",
          "description": "",
          "opacity": 0.8,
          "defaultStyle": {
            "regionColumn": "GEOID",
            "color": {
              "numberOfBins": 5,
              "colorPalette": "Oranges"
            }
          }
      }
    ],
    "viewerMode": "3dSmooth",
    "baseMaps": {
        "defaultBaseMapId": "basemap-positron",
        "previewBaseMapId": "basemap-positron"
    }
}

please check my data file, regionId, and mvt files:
[data.zip](https://github.com/TerriaJS/terriajs/files/9099464/data.zip)

It would be great appreciate to help me.

**Reply by Joebjammin** [reply]
@nf-s Nick, Could I trouble you to put ALL the files in a zip file, indicating the proper path. Three of us have been trying to get it to work to no avail. Oh yes, please indicate the version of Terria that you tested on. Once we get this to work, I can help with new documentation, if you like.

**Reply by nf-s** [reply]
Hi @Joebjammin 

Before I do that, are you able to seen any failed network requests in your browser console?
Also, which `regionMapping.json` file are you editing? 

The default file loaded by TerriaMap is `build/TerriaJS/data/regionMapping.json`. This can be configured by setting `regionMappingDefinitionsUrl` in your `wwwroot/config.json`

You can confirm that the correct one is being loaded by inspecting it in the "Network" tab in your browser console

<img width="1281" alt="image" src="https://user-images.githubusercontent.com/6187649/179481054-b2328391-0610-47f2-ab5d-fb7a04af2bc6.png">

Both the `regionMapping.json` and `data/regionids/region_map-FL_COUNTY.json` files should be loaded when you add the CSV to the map.

Cheers,
Nick

**Reply by nf-s** [reply]
Also I would have been using TerriaJS `8.2.8`

**Reply by Joebjammin** [reply]
Hi Nick, Thanks for your patience. You revealed the issue in your answer. We assumed that if we just replaced the contents of the regionMapping file we were golden, thinking that region mapping process was changed in v 8. In v 7 config, this option was commented out and alerted me to change. Once I changed "regionMappingDefinitionsUrl": "data/regionMapping.json" .... voila Terria magic. 


You guidance is much appreciated. Have a good one. Love this application. Regards, Joe

**Reply by nf-s** [reply]
Hi @Joebjammin 

That's great, no problem at all!
Sorry the documentation is a bit all over the place at the moment - and there are a few remnants of v7 still around.

Cheers,
Nick
